### ISSUE TITLE
Groups is empty

### ISSUE THREAD

**Issue by joakim-hove** [snok/django-auth-adfs#325]
Hello;

I have used `django-auth-adfs` to log in users on Azure/Entra. The users are correctly authenticated, but unfortunately the list of groups received from Azure is empty. I need the group memberships for my application to work correctly - so unfortunately I am not fully sucesfull yet. With some of my own print() debug in `djangu-auth-adfs` this is the dialogue between my server and Azure:
``` 
[18/Jan/2024 12:51:54] "GET /oauth2/login?next=/ HTTP/1.1" 302 0
Groups: []              <--- My debug
User: user@domain.com   <--- My debug
[18/Jan/2024 12:51:54] "GET /oauth2/callback?code=0.AYIAS3tqt2........ccbe62f15376 HTTP/1.1" 302 0
[18/Jan/2024 12:51:54] "GET / HTTP/1.1" 200 12260
```

as indicated in the console log the groups list is empty.

I have given the application the permission: `GroupMember.Read.All` - but apart from that I am not very experienced in Azure/Entra configuration - so the problem might be on that side. Any tips would be highly appreciated.

This is my settings section:
```
AUTH_ADFS = {
    'AUDIENCE': ENTRA_CLIENT_ID,
    'CLIENT_ID': ENTRA_CLIENT_ID,
    'CLIENT_SECRET': ENTRA_SECRET,
    'CLAIM_MAPPING': {'first_name': 'given_name',
                      'last_name': 'family_name',
                      'email': 'upn'},
    'GROUPS_CLAIM': 'groups',
    'MIRROR_GROUPS': True,
    'USERNAME_CLAIM': 'upn',
    'TENANT_ID': ENTRA_TENANT_ID,
    'RELYING_PARTY_ID': ENTRA_CLIENT_ID,
    "GROUP_TO_FLAG_MAPPING": {"is_staff": ["SomeGroup"],
                              "is_superuser": ["SomeGroup"]},
}
```

## Upvote & Fund

- We're using [Polar.sh](https://polar.sh/snok) so you can upvote and help fund this issue.
- We receive the funding once the issue is completed & confirmed by you.
- Thank you in advance for helping prioritize & fund our backlog.

<a href="https://polar.sh/snok/django-auth-adfs/issues/325">
<picture>
  <source media="(prefers-color-scheme: dark)" srcset="https://polar.sh/api/github/snok/django-auth-adfs/issues/325/pledge.svg?darkmode=1">
  <img alt="Fund with Polar" src="https://polar.sh/api/github/snok/django-auth-adfs/issues/325/pledge.svg">
</picture>
</a>

**Comment by tim-schilling** [comment]
Have you looked at what AD is sending back in its response?

**Comment by joakim-hove** [comment]
I am sorry to be so late at getting back to this - I have offline travelling. Will be online from now. 

> Have you looked at what AD is sending back in its response?

Yes - I have looked at the response AD is sending back. The very basic raw communication in the shell looks like this (I have removed most of the content - happy to share in full through another channel).
```
[27/Jan/2024 07:28:53] "GET /oauth2/callback?code=0.AYIAS3tqt2qkVkqhdXG[.......] 89G_TZjz8l1vI8AdxQqNP4EzuWvAo&state=Lw%3d%3d&session_state=[....] HTTP/1.1" 302 0
```

I have added some manual `print(...)` debugging in the `django-auth-adfs` source code, which in prints the following in the console:
```
Groups: []
User: user@xxx.com
```
The result of 
```
for key, value in claims.items():
     print(f"{key} = {value})
```
looks like this (NB - I have replaced the value with `xxxx`except for the `scp` field) - happy to provide details in other channel

aud = xxx
iss = https://sts.windows.net/xxx
iat = xxxx
nbf = xxx
exp = xxx
acr = xxx
aio = xxxx
amr = ['pwd', 'mfa']
appid = xxxx
appidacr = 1
family_name = xx
given_name = xxx
ipaddr = xxxx
name = xxxx
oid = xxxx
rh = xxxx
scp = email Group.Read.All GroupMember.Read.All offline_access openid PrivilegedAccess.Read.AzureADGroup PrivilegedAssignmentSchedule.Read.AzureADGroup PrivilegedEligibilitySchedule.Read.AzureADGroup profile RoleManagementPolicy.Read.AzureADGroup User.Read
sub = xxxx
tid = bxxxx
unique_name = xxxx
upn = xxxx
uti = xxx
ver = 1.0
```

I have tried to use the "Fund this issue" functionality, but it was unclear to me whether that actually worked. Please indicate to me if the "Fund this issue" has not kicked in, and I will try again.

**Comment by JonasKs** [comment]
No worries. Thank you for using the funding button, it really makes a difference for open source people like us😊 It worked! 

Your token doesn't contain the `groups` claim which it is looking for, based on your settings: `'GROUPS_CLAIM': 'groups',`. 

Is your intention to use `scopes` over group memberships? Or combined?   

Could you show your settings for the Group claims in Azure? It should be under the configuration for access tokens. If not, I'll check tomorrow when I'm on a laptop. 
I think you're supposed to select the `sAMAccountName`. 

![pic](https://user-images.githubusercontent.com/593327/129056240-e72b6d0f-b95c-4237-9e7e-2eba009c7baa.png)

**Comment by JonasKs** [comment]
As a side note - groups are global groups for your entire org. 

It's often better to create roles specifically managed by your appreg. I can explain this more in detail if you'd like.

**Comment by joakim-hove** [comment]
> No worries. Thank you for using the funding button, it really makes a difference for open source people like us😊 It worked!

Good!

> 
> Your token doesn't contain the `groups` claim which it is looking for, based on your settings: `'GROUPS_CLAIM': 'groups',`.

That agrees with my understanding of the problem.

> 
> Is your intention to use `scopes` over group memberships? Or combined?

I am sorry to say that I do not really understand the question. I am a (fairly) experienced Django/Python developer - but user administration with Azure is a completely new and opaque topic to me. My intention was as follows (at least for the start):

1. Create a group in Azure AD (that I have done).
2. Assign selected members of my organization to this this group in Azure AD.
3. When I log in with `django-auth-adfs` the token should contain a list of groups the user is part of, and then the Django application should behave accordingly. This is where I struggle - because that groups claim is empty. 

> 
> Could you show your settings for the Group claims in Azure? It should be under the configuration for access tokens. If not, I'll check tomorrow when I'm on a laptop. I think you're supposed to select the `sAMAccountName`.

This is the token configuration for my app:
![image](https://github.com/snok/django-auth-adfs/assets/1321665/d0b5bb19-0def-4de3-be72-d332f05b4af0)

i.e. empty - that might be the problem?!

> 
> ![pic](https://user-images.githubusercontent.com/593327/129056240-e72b6d0f-b95c-4237-9e7e-2eba009c7baa.png)

**Comment by JonasKs** [comment]
I'd defined use app roles if you have your own group for this app. 😊  
Makes it easier to manage, and your token only need to give roles associated with your app, not all the roles in your entire organization. 

I'll give you a step by step tomorrow.

**Comment by joakim-hove** [comment]
I then tried to add a "Groups claim" as this:

![image](https://github.com/snok/django-auth-adfs/assets/1321665/b5000902-4814-4f38-8b40-2d7857001e49)

Now the payload actually has a groups list - but it is in the form of a list of uuids:
```
groups = ['66fe5b4e-de2f-4369-8617-84c464ab634e', '4f131d51-4993-4201-8350-5a35ac11a5a6', '41a4f6cc-51ff-4125-914a-37e42281712b', '83f714e4-fec6-4370-831f-11932d856cf4']
....
wids = ['62e90394-69f5-4237-9190-012177145e10', 'b79fbf4d-3ef9-4689-8143-76b194e85509']
```
I guess that might be progress?

**Comment by joakim-hove** [comment]
> I'll give you a step by step tomorrow.

That'd be wonderfull :smile:

**Comment by JonasKs** [comment]
Yeah, this is the entire orgs groups. There's no name on them if they're not associated to ADFS I think..

If you go into your appreg overview, then click the enterprise registration, you can create a custom role there, only for this appreg. Then you can assign users to it.

**Comment by joakim-hove** [comment]
> If you go into your appreg overview, then click the enterprise registration, you can create a custom role there, only for this appreg. Then you can assign users to it.

Lo and behold - with good guidance and extensive amount of random clicking in the Azure AD portal I have managed to get this into the venerable "Works for me" state - now I'll "just" have to check that it works for my colleagues on monday :crossed_fingers:

**Comment by JonasKs** [comment]
Awesome 😊 Let us know. Thanks again for funding the issue!

**Comment by JonasKs** [comment]
@joakim-hove , did you figure this out? 😊

**Comment by joakim-hove** [comment]
No - I have not got around to verifying it. But it works for me - so let's assume it works for my colleagues as well. You can consider this issue closed and accept the funds.

**Comment by joakim-hove** [comment]
Hello;

I have now tested this with my colleagues - and it works for them as well - when running on localhost. But - unfortunetaly it does not work in production. In production I get an error pop in the browser with the message: 
```
AADSTS50011: The redirect URI 'http://sonairmeasurements.azurewebsites.net/oauth2/callback' specified in the 
request does not match the redirect URIs configured for the application
```
the URI quoted in the error message is with plain 'http://' - whereas the callback I have configured in Azure is with `https://` - so in a sense the error message from Azure is correct - if the payload from my server indeed contains the plain `http://`?
